tests/loops/test_loops.py

```diff
@@ -57,18 +77,145 @@ def load_state_dict(self, state_dict: Dict) -> None:
+            self.increment = state_dict["increment"]
+
+    class Simple(Loop):
+
+        def __init__(self, a):
+            super().__init__()
+            self.a = a
+            self.progress = SimpleProgress()
+
+        def advance(self, *args: Any, **kwargs: Any) -> None:
+            loop = getattr(self, "loop_child", None)
+            if not loop:
+                return
+            loop.run()
+            self.progress.increment += 1
+
+        @property
+        def skip(self) -> bool:
+            return False
+
+        @property
+        def done(self) -> bool:
+            return self.iteration_count > 0
+
+        def reset(self) -> None:
+            self.restarting = False
+
+        def on_save_checkpoint(self) -> Dict:
+            return {"a": self.a}
+
+        def on_load_checkpoint(self, state_dict: Dict) -> None:
+            self.a = state_dict["a"]
+
+    grand_loop_parent = Simple(0)
+    loop_parent = Simple(1)
+    loop_child = Simple(2)
+
+    loop_parent.loop_child = loop_child
+
+    assert not loop_parent.skip
+
+    state_dict = loop_parent.state_dict()
+
+    loop_progress = _collect_loop_progress(loop_parent)
+    assert loop_progress["progress"] == loop_parent.progress
+    assert loop_progress["loop_child"]["progress"] == loop_child.progress
+
+    loop_progress = _collect_loop_progress(loop_child)
+    assert loop_progress["progress"] == loop_child.progress
+
+    loop_parent.trainer = Trainer()
+    assert loop_child.trainer == loop_parent.trainer
+
+    assert state_dict == OrderedDict([('state_dict', {
+        'a': 1
+    }), ('progress', {
+        'increment': 0
+    }), ('loop_child.state_dict', {
+        'a': 2
+    }), ('loop_child.progress', {
+        'increment': 0
+    })])
+
+    loop_parent.progress
+
+    state_dict["loop_child.state_dict"]["a"] = 3
+
+    loop_parent.load_state_dict(state_dict)
+    assert loop_parent.restarting
+
+    loop_parent.run()
+
+    loop_parent_copy = deepcopy(loop_parent)
+    assert loop_parent_copy.state_dict() == loop_parent.state_dict()
+
+    assert loop_parent_copy.on_save_checkpoint() == {'a': 1}
+    assert loop_parent_copy.loop_child.on_save_checkpoint() == {'a': 3}
+
+    assert not loop_parent.restarting
+
+    state_dict = loop_parent.state_dict()
+    assert state_dict == OrderedDict([('state_dict', {
+        'a': 1
+    }), ('progress', {
+        'increment': 1
+    }), ('loop_child.state_dict', {
+        'a': 3
+    }), ('loop_child.progress', {
+        'increment': 0
+    })])
+
+    loop_parent = Simple(1)
+    loop_child = Simple(2)
+    loop_parent.loop_child = loop_child
+    loop_parent.load_state_dict(state_dict)
+    assert loop_parent.progress.increment == 1
+    assert loop_parent.loop_child.progress.increment == 0
+
+    del loop_parent.loop_child
+    state_dict = loop_parent.state_dict()
+    assert state_dict == OrderedDict([('state_dict', {'a': 1}), ('progress', {'increment': 1})])
```

Comment:
The state dicts of the Loop class do not return `OrderedDict`s. Should they?

If not, why are we testing them here?

Reply:
Good question. Not anymore. This test should be updated.